### ISSUE TITLE
Feature Idea: Ballot Critical Path Analysis

### ISSUE THREAD

**Issue by philipbelesky** [czlee/tabbycat#392]
So while the current dashboard graph showing number of ballots over time is fine, it doesn't actually help give insight into how smoothly a round went.

I think an ideal graph would show all the ballots in a form that allows for easy identification of:

- When the very first ballot came in
- The spans of time between when each ballot came in and was finalised
- Which ballots are furthest behind in the process
- The broad 'shape' of the critical path: are all ballots being entered too slowly? Are there particular ballots that cause exceptions? How long is the turn around time from a data entry perspective?

This would look sort of like [a box plot](http://bl.ocks.org/mbostock/4061502) / [stacked bar chart](http://bl.ocks.org/wpoely86/e285b8e4c7b84710e463) with a timescale and hoverable details. Each line would be a ballot.

It also strikes me that some tournaments might want this to be public, for transparency's sake. So maybe we should have an option for that.

**Comment by czlee** [comment]
When I had this idea in mind I was also thinking of other parts of the critical path of a tournament, for example, tracking when debates start, and how long generating the draw, allocating adjudicators and rolling the draw in GA takes. That said, it's not clear to me that all of these are well-defined, or can be easily tracked automatically.

To the extent that some things can be tracked automatically, they already are in the action log, so it might just be a case of analyzing the data that's already there. That's the case with all of the ballot entry stages, for example.

It'd also have to account for the case where some ballots aren't checked in before they're entered. We wouldn't want this system itself to cause any overhead.

Would the public display extend to naming those debates (and hence panels) that take the longest?

**Comment by philipbelesky** [comment]
Totally — I think the graph/diagram should include as much of the timelines as can be reliably determined.

For the public display, at the point people are willing to turn it on it's probably best to include identifiers for panels. If just to reduce the UI complexity, but I guess any self-policing/shaming effects could be useful as well.

**Comment by philipbelesky** [comment]
In light of the new ballots graph this can probably take on a much more simple form: I'm thinking just a horizontal line graph represented like a transport network diagram (ie dots and lines, similar to how the sidebar works). Basically it would indicate the current time (at far right; unless the round was advanced in which case that is the timeline's end) as a pip with earlier pips for:

- Time the round was advanced to
- Time the draw was confirmed
- Time the first adj allocation was made
- Time the last adj allocation was made
- Time the draw was released
- Time the motions were released
- Time that debates started
- Time the first ballot was entered
- Time the last ballot was entered

Some problems with scale: we might want a maximum time length — ie if you advanced the round last night but only did the draw this morning, that stretch of the graph shouldn't then be huge. Ideally it would show some aspect of truncation if the time gap is greater than an hour.

Unclear whether this should be in d3 or just some fancy css.

**Comment by czlee** [comment]
I agree, and I think there are different "levels" of detail. The pip-line above would be the macro-level, and at the micro-level you have individual ballots and the like.

On another note: we'd want to include an ability to change the round being viewed in the UI. Currently the ballots graph is always for the current round, which is good on the tournament overview page, but for diagnostics we want to be able to look at previous rounds.

**Comment by czlee** [comment]
Also, we should add ballot check-ins because they can gave insight into whether delays were in ballots coming back to the tab room or data entry being slow to get to ballots that have arrived, but we shouldn't assume that ballot check-ins are being used.

**Comment by philipbelesky** [comment]
Agree that there should be a way to view earlier rounds. I assume this is the best way of doing it:

- [ ] Create a tournament-overview page for a specific round and make it accessible from the Results view

With regard to checkins, shouldn't ballot checkins be in the stacked area graph rather than the line graph? The same goes with other micro-level details; that stacked area graph seems like the most comprehensive record. Although that said it's not hard to just go to the Results page and check the status of individual ballots.

**Comment by philipbelesky** [comment]
As highlighted in 89d0f14875deb837e673db51e838782e5e6e0eca:

- [ ] The responsive width isn't really working properly, on my browser the left edge is fine but the right edge goes past the container.

**Comment by czlee** [comment]
Oh, it might not need to be a tournament overview page per se, I don't imagine "latest results" and "latest actions" are very useful on a previous round page. I guess this would be a page in its own right, a subset of which is also on the overview page? Though I guess I can't think of anything in the critical path analysis that I also wouldn't like to see at-a-glance on the tab room projector.

Re check-ins, I agree. The only thing preventing me from putting it in the current graph was that its data is in a different format—check-in timestamps are only in the action logs, so either we'd need to add a check-in timestamp to the Debate model, or we'd need to get the statistics code to interpret action logs for this. Ideally it would also show which ballot a step relates to when you roll over it 😄 though that's a little overambitious for me.

**Comment by tienne-B** [comment]
A line or area chart would be much more appropriate for this visualization, with different series for the different statuses of ballots. In all, it would always show a total of 100%, but when _au fur et à mesure ballots are being treated, the series changes to reflect that.

**Comment by philipbelesky** [comment]
Isn't that the function that the ballots status graph provides on the tournament dashboard/home page?

The intent here is slightly different in that we only care about the start/end of data entry as one of several critical time intervals when running a round. E.g. how long the adj allocation takes or the gap between allocation and motion release (as an example the kind of data underpinning [the table shown under Scheduling on this page](https://tabbycat.readthedocs.io/en/stable/guide/tournament-logistics.html). So the visual imagined here is less a chart than a simple timeline.
